docs/encyclopedia/application-message-passing.mdx

```diff
@@ -139,76 +197,161 @@ Stack Trace Queries are available only for running Workflow Executions.
-   In this case, the system rejects the request, doesn't record anything in the Workflow Event History to indicate that the Update ever happened, and the Update processing doesn't proceed to later phases.
-   If the Update completes the validation stage without error, the Platform changes its state to **Accepted** and a [WorkflowExecutionUpdateAcceptedEvent](/references/events#workflowexecutionupdateacceptedevent) Event is added to Workflow Execution [Event History](/workflows#event-history).
-3. **Execution.** Accepted Update requests move to the execution phase.
-   In this phase, the Worker delivers the request to the Update handler.
-4. **Completion.** The Update handler can return a result or a language-appropriate error/exception to indicate its completion.
-   The Platform sends the Update outcome back to the original invoking entity as an Update response.
-   A [WorkflowExecutionUpdateCompletedEvent](/references/events#workflowexecutionupdatecompletedevent) Event in the Workflow Execution Event History denotes the completion of an Update.
+- Throw any exception from your validator, OR
+- Throw an [Application Failure](/references/failures#application-failure) from your Update handler. Unlike with Signals, the Workflow will keep going.
 
-:::note
+If you throw any other exception, by default, it will cause a [Workflow Task Failure](references/failures#workflow-task-failures). This means the Workflow will get stuck and will retry the handler periodically until the exception is fixed, for example by a code change or infrastructure coming back online. Note that this will cause a delay for clients waiting for a result.
 
-Workflow Updates are currently disabled by default on Temporal Server.
+#### Exceptions in message handlers in the Go SDK
 
-To enable the `UpdateWorkflowExecution` API, set the [frontend.enableUpdateWorkflowExecution](https://github.com/temporalio/temporal/blob/main/common/dynamicconfig/constants.go) dynamic config value to `true`.
+In Go, most exceptions behave like [Application Failures](/references/failures#application-failure) in the other SDKs. Panics behave like other exceptions and fail the Signal or Update handler tasks.
 
-For example, to enable Workflow Updates with the Temporal CLI, pass the value when executing the Temporal CLI command:
+#### Idempotency
 
-```command
-temporal server start-dev --dynamic-config-value frontend.enableUpdateWorkflowExecution=true
-```
+Most developers want their message handlers to run exactly once in cases where a Signal is delivered twice or sent by two different call sites. You can pass a unique key among your message arguments and use that to check for duplicate messages in your handler.
+For Updates, clients can instead pass the ID argument into the Update to serve as an idempotency key and as logging context. This can be retrieved within your handler and used to check whether the Update has already been handled.
```

Comment:
I think that some of this idempotency advice is superfluous.
- Updates are deduplicated by the server on updateID, but this section makes that fact unclear.
- For Signals there is a "request ID" deduplication mechanism that I think covers some subset of the "delivered twice" error mode
- I think it would help to clarify what it means for a signal to be delivered redundantly by two different call sites. I suspect that this concept is confusing to people, since of course Temporal absolutely allows multiple invocations of the same signal with the same arguments around the same time to be considered non-redundant.

Reply:
Our deduplication does not work across CaN and in my mind that's too much of a gotcha to be able to recommend anything other than good hygiene here.  (we could say "if your workflow doesn't CaN" but then people will copy/paste or the workflow will add CaN later.)
Let's chat offline if we want to keep thinking here.

Reply:
Re: signal, at least in python, looks like the signal's `request_id` is not exposed and is set to a uuid.  So no ability for a client to combine different callsites and I'm not sure it's worth recommending.  But maybe we should open tasks to make it consistent with updates?

Reply:
Okay, I decided to be more verbose about it because if people try to test their deduplication, they may run into trouble without fully understanding what's going on.